File: app/sepuling-java/src/main/java/smecalculus/bezmen/storage/SepulkaDaoMyBatis.java
package smecalculus.bezmen.storage;

import java.util.Optional;
import java.util.UUID;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import smecalculus.bezmen.core.SepulkaStateDm;
import smecalculus.bezmen.storage.mybatis.SepulkaSqlMapper;

@RequiredArgsConstructor
public class SepulkaDaoMyBatis implements SepulkaDao {

    @NonNull
    private SepulkaStateMapper stateMapper;

    @NonNull
    private SepulkaSqlMapper sqlMapper;

    @Override
    public SepulkaStateDm.AggregateRoot addNew(@NonNull SepulkaStateDm.AggregateRoot state) {
        var stateEdge = stateMapper.toEdge(state);
        sqlMapper.insert(stateEdge);
        return state;
    }

    @Override
    public Optional<SepulkaStateDm.Existence> getBy(@NonNull String externalId) {
        return sqlMapper.findByExternalId(externalId).map(stateMapper::toDomain);
    }

    @Override
    public Optional<SepulkaStateDm.Viewing> getBy(@NonNull UUID internalId) {
        return sqlMapper.findByInternalId(internalId).map(stateMapper::toDomain);
    }

    @Override
    public void touchBy(@NonNull UUID internalId, @NonNull SepulkaStateDm.Touch state) {
        var stateEdge = stateMapper.toEdge(state);
        var matchedCount = sqlMapper.updateBy(internalId, stateEdge);
        if (matchedCount == 0) {
            throw new ContentionException();
        }
    }
}

File: app/sepuling-kotlin/src/main/kotlin/smecalculus/bezmen/storage/SepulkaDaoMyBatis.kt
package smecalculus.bezmen.storage

import smecalculus.bezmen.core.SepulkaStateDm
import smecalculus.bezmen.storage.mybatis.SepulkaSqlMapper
import java.util.UUID

class SepulkaDaoMyBatis(
    private val stateMapper: SepulkaStateMapper,
    private val sqlMapper: SepulkaSqlMapper,
) : SepulkaDao {
    override fun addNew(state: SepulkaStateDm.AggregateRoot): SepulkaStateDm.AggregateRoot {
        val stateEdge = stateMapper.toEdge(state)
        sqlMapper.insert(stateEdge)
        return state
    }

    override fun getBy(externalId: String): SepulkaStateDm.Existence? {
        return sqlMapper.findByExternalId(externalId)?.let { stateMapper.toDomain(it) }
    }

    override fun getBy(internalId: UUID): SepulkaStateDm.Viewing? {
        return sqlMapper.findByInternalId(internalId)?.let { stateMapper.toDomain(it) }
    }

    override fun touchBy(
        internalId: UUID,
        state: SepulkaStateDm.Touch,
    ) {
        val stateEdge = stateMapper.toEdge(state)
        val matchedCount = sqlMapper.updateBy(internalId, stateEdge)
        if (matchedCount == 0) {
            throw ContentionException()
        }
    }
}
